### ISSUE TITLE
Standard credential handling workflow

### ISSUE THREAD

**Issue by mih** [datalad/datalad-next#330]
We don't have that yet, we only have (various) implementations. `CredentialManager` has `obtain()`, but this is only covering some aspects. We need:

- a reliable and flexible way to declare credentials for non-interactive operation
- we need a convenient and frustration minimizing approach to (manual) credential selection, entry, retry for interactive use. In particular the *retry* part is largely unaddressed or custom right now.

#### Scenario 1: non-interactive

- a required credential must be known before
- a resource that requires a credential that is *not* known upfront, must not be tried, or skipped with the least possible costs. No attempt should be made to prompt for a credential, even when a known credential did not enable access

#### Scenario 2: interactive, credential unknown

- user must be informed that a credential is needed, and be given the ability to enter one
- ideally, a user would not need to stop and redo the operation, if the credential is in fact known, but simply not discoverable (missing realm, whatever), but could also just provide the credential name
- when a credential does not work (mistyped, etc), a user should have the ability to retry (without having to start from scratch)
- when a user realizes that they in fact have no adequate credentials, they should have to ability to say "I don't have any", without having to kill the entire operation

#### Scenario 2: interactive, credential known

- when a known credential fails to enable access, a user should be given the opportunity to select another credential, to edit a previously selected or discovered credential, and try again
- when a user realizes that they cannot fix a credential, they should be given the option to "proceed" without a credential (IOW skip the credential entry, and live with the consequences)

**Comment by adswa** [comment]
The scenarios that you describe make sense to me, and cover real usecases. Allow me a question to guide my involvement here: What are you looking for at this stage as feedback - more usecases/scenarios, feedback on the ones you outline, implementation discussions, or all of those things/something else entirely?

**Comment by mih** [comment]
@adswa All of those. We need to 

- make sure we understand the problem space completely
- come up with solutions
- check if the can replace the partial solution that are implemented already

re solutions:

My thinking currently considers extending the workflow implementation that are already covered by `CredentialManager.obtain()`. It presently implements a one-time-retrieval, but it is ignorant of a scenario where whatever it retrieved is insufficient to meet the goal of the retriever/consumer of a credential.

I am considering whether the API of `obtain()` could be extended to receive a a custom callable (provided by the caller) that `obtain()` calls with the retrieved credential, and it returns a boolean that indicates whether this credential is fit for the intended purpose (e.g., login worked).

With that, we could make `obtain()` try more than once, and also become the single place to implement user interactions (`CredentialManager` already does things like asking-to-save-under-a-given-name).

**Comment by adswa** [comment]
Looking into how (some) of the downloaders framework works, would it alternatively be feasible to wrap ``obtain()`` in a loop for the interactive scenarios to give users control over the retry? something like:

```
if interactive:
    repeat = True
    while repeat:
        cred = obtain(...)
        if not cred:
              repeat = ui.question("Wanna retry?")
```

**Comment by mih** [comment]
@adswa Yes, that is possible. But to me the question is whether that loop could/should not be in `obtain()` and be conditional *not* upon the ability to retrieve a credential, but upon explicit feedback (from a caller provided tester) on whether the retrieved credential is fit-for-purpose.

If the loop is outside `obtain()`, either all consuming code needs to implement this (separately), or we provide a standard wrapper around `obtain()` -- which is also fine, but raises the question what consuming code should *not* use that wrapper instead of `obtain()`?

**Comment by adswa** [comment]
True. I think it makes sense to have the possibility to retry credentials entry everywhere in a consistent manner. The fewer wrappers need to be remembered to be used, the more consistent will behavior be across commands.

**Comment by mih** [comment]
Just stumbled upon another bad UX case:

I have a credential for sciebo stored. I am accessing a share that is on sciebo, but under a different account. The auth code sees that I have a matching credential, supplies it, fails, and fails the whole thing, because it is simply the wrong credential. All interactive, no questions, plain error-exit.

I would have expected:

"I tried access with known credential <name>, but it failed. Would you like to provide a different credential?"

**Comment by christian-monch** [comment]
> But to me the question is whether that loop could/should not be inside obtain()

As you pointed out, we have to run two loops in this scenario, i.e. resource access on one hand and credential provision on the other hand. We can delegate the resource access validation to `obtain()`. It feels somehow not optimal, but I have a small example program that fulfills all requirements (as far as I can see):

<details><summary>a small demo</summary>
```python
from __future__ import annotations

from dataclasses import dataclass
from typing import Iterable


@dataclass
class Credential:
    name: str
    user_id: str
    resource_id: str
    credential: str


class CredentialManager:
    def __init__(self,
                 initial_credentials: Iterable[Credential] | None = None):
        self.credentials = {
            credential.name: credential for credential in tuple(initial_credentials or [])
        }

    def get_credentials(self, user_id: str, resource_id: str) -> list[Credential]:
        return [cred for cred in self.credentials.values() if cred.user_id == user_id and cred.resource_id == resource_id]

    def get_named_credential(self, name: str) -> Credential | None:
        return self.credentials.get(name, None)

    def add_credential(self, credential: Credential):
        self.credentials[credential.name] = credential


cm = CredentialManager([
    Credential(name='a@1', user_id='u_a', resource_id='res_1', credential='pass'),
    Credential(name='a@2', user_id='u_a', resource_id='res_2', credential='pass'),
    Credential(name='a@3', user_id='u_a', resource_id='res_3', credential='no_pass'),
    Credential(name='b@1', user_id='u_b', resource_id='res_1', credential='pass'),
    Credential(name='b@2.1', user_id='u_b', resource_id='res_2', credential='no_pass'),
    Credential(name='b@2.2', user_id='u_b', resource_id='res_2', credential='pass'),
    Credential(name='b@2.3', user_id='u_b', resource_id='res_2', credential='no_pass')
])


def validate_access(user_id: str, resource_id: str, credential: Credential):
    return credential.credential == 'pass'


def input_credential(resource_id, validate_cb) -> list[Credential] | None:
    """Get a credential or a "no credential" info from the user
    
    This method asks the user to enter a credential. The user can
    either:
    
    - enter a new credential
    - signal that he does not want to enter a credential ("!x")
    - refer to an existing credential ("!n <credential.name>")
    
    The input loop is repeated until a credential or an exit-wish
    is provided. 
    """
    while True:
        print(f'Enter credential for resource {resource_id} ("<credential name>:<user>:<password>" or "!x" to not access resource "!n <credential name>" for an existing credential')
        user_input = input('> ')
        if user_input == '!x':
            return None
        elif user_input.startswith('!n '):
            credential_name = user_input[3:]
            credential = cm.get_named_credential(credential_name)
            if credential is None:
                print(f'[I: UNKNOWN CRED NAME] no credential with name: {credential_name} ')
                continue
        else:
            name, user_id, password = user_input.split(':')
            credential = Credential(name, user_id, resource_id, password)

        if validate_cb(credential.user_id, credential.resource_id, credential):
            cm.add_credential(credential)
            return [credential]
        print('[I: INVALID CRED ENTERED')

def obtain(user_id: str, resource_id: str, is_interactive: bool, validate_cb) -> Credential | None:
    credentials = cm.get_credentials(user_id, resource_id)
    if not credentials:
        print(f'[N/I: NO CRED STORED] no credential available for {user_id} at {resource_id}')
        if is_interactive:
            credentials = input_credential(resource_id, validate_cb)
            if not credentials:
                print(f'[I: NO VALID CRED PROVIDED] {user_id} skipping {resource_id} because no valid credential was provided')
                return None
        else:
            return None

    for credential in credentials:
        if validate_cb(user_id, resource_id, credential):
            print(f'[N/I: GOOD CRED] {user_id} is accessing {resource_id} with credential {credential}')
            return credential
        else:
            print(f'[N/I: BAD CRED] {user_id} has bad credential {credential} for {resource_id}')

    if is_interactive:
        print(f'[I: NO VALID CRED STORED] No valid credential available for {user_id} at {resource_id}')
        credentials = input_credential(resource_id, validate_cb)
        if not credentials:
            print(f'[I: NO VALID CRED PROVIDED] no valid credential for {resource_id} was provided')
            return None
        return credentials[0]
    print(f'[N: NO VALID CRED] {user_id} has no valid credential for {resource_id}')
    return None
    

def credential_demo(user_id: str, resource_ids: list[str], is_interactive, validate_cb):
    for resource_id in resource_ids:
        credential = obtain(user_id, resource_id, is_interactive, validate_cb)
        if credential is None:
            print(f'[N/I: NO ACCESS] no valid credential for {resource_id}, skipping it')


credential_demo('u_a', ['res_1', 'res_2', 'res_3'], False, validate_access)
credential_demo('u_b', ['res_1', 'res_2', 'res_3'], False, validate_access)
credential_demo('u_c', ['res_1', 'res_2', 'res_3'], False, validate_access)


credential_demo('u_a', ['res_1', 'res_2', 'res_3'], True, validate_access)
credential_demo('u_b', ['res_1', 'res_2', 'res_3'], True, validate_access)
credential_demo('u_c', ['res_1', 'res_2', 'res_3'], True, validate_access)
``` 
</details>

It is not too elegant (what is, actually? ;-) ) I am not sure about the delegation of the validation into `obtain()`, I have a version that keeps validation out of the credential-code path. But it is more redundant.

**Comment by christian-monch** [comment]
This would be a simple example with delegation of credential validation to `obtain()`:

<details> <summary> small demo with delegation </summary>

```python

from __future__ import annotations

from dataclasses import dataclass
from typing import Iterable


@dataclass
class Credential:
    name: str
    resource_id: str
    credential: str


class CredentialManager:
    def __init__(self,
                 initial_credentials: Iterable[Credential] | None = None):
        self.credentials = {
            credential.name: credential for credential in tuple(initial_credentials or [])
        }

    def get_credentials(self, resource_id: str) -> list[Credential]:
        return [cred for cred in self.credentials.values() if cred.resource_id == resource_id]

    def get_named_credential(self, name: str) -> Credential | None:
        return self.credentials.get(name, None)

    def add_credential(self, credential: Credential):
        self.credentials[credential.name] = credential


# Initialize the credential manager with:
# - a valid credential for resource 'res_1'
# - an invalid credential for resource 'res_2'
# - no credential for resource 'res_3'
cm = CredentialManager([
    Credential(name='ok@1', resource_id='res_1', credential='pass'),
    Credential(name='bad@2', resource_id='res_2', credential='no_pass'),
])


# A simple validator, all credential infos where `credential` equals `'pass'` are valid.
def validate_access(resource_id: str, credential: Credential):
    return credential.credential == 'pass'


def input_credential(resource_id, validate_cb) -> list[Credential] | None:
    """Get a credential or a "no credential" info from the user
    
    This method asks the user to enter a credential. The user can
    either:
    
    - enter a new credential (format: "<credential name>:<credential content>")
    - signal that he does not want to enter a credential ("!x")
    - refer to an existing credential ("!n <credential name>")
    
    The input loop is repeated until a credential or an exit-wish
    is provided. 
    """
    while True:
        print(f'Enter credential for resource {resource_id} ("<credential name>:<credential content>") or "!x" to not access resource "!n <credential name>" for an existing credential')
        user_input = input('> ')
        if user_input == '!x':
            return None
        elif user_input.startswith('!n '):
            credential_name = user_input[3:]
            credential = cm.get_named_credential(credential_name)
            if credential is None:
                print(f'[I: UNKNOWN CRED NAME] no credential with name: {credential_name} ')
                continue
        else:
            name, content = user_input.split(':')
            credential = Credential(name, resource_id, content)

        if validate_cb(credential.resource_id, credential):
            cm.add_credential(credential)
            return [credential]
        print('[I: INVALID CRED ENTERED')


def obtain(resource_id: str, is_interactive: bool, validate_cb) -> Credential | None:
    mark = 'I' if is_interactive else 'N'
    credentials = cm.get_credentials(resource_id)
    if not credentials:
        print(f'[{mark}: NO CRED STORED] no credential available for {resource_id}')
        if is_interactive:
            credentials = input_credential(resource_id, validate_cb)
            if not credentials:
                print(f'[{mark}: NO VALID CRED PROVIDED] skipping {resource_id} because no valid credential was provided')
                return None
        else:
            return None

    for credential in credentials:
        if validate_cb(resource_id, credential):
            print(f'[{mark}: GOOD CRED] accessing {resource_id} with credential {credential}')
            return credential
        else:
            print(f'[{mark}: BAD CRED] bad credential {credential} for {resource_id}')

    if is_interactive:
        print(f'[{mark}: NO VALID CRED STORED] No valid credential available for {resource_id}')
        credentials = input_credential(resource_id, validate_cb)
        if not credentials:
            print(f'[{mark}: NO VALID CRED PROVIDED] no valid credential for {resource_id} was provided')
            return None
        return credentials[0]
    print(f'[{mark}: NO VALID CRED] no valid credential for {resource_id}')
    return None
    

def credential_demo(resource_ids: list[str], is_interactive, validate_cb):
    mark = 'I' if is_interactive else 'N'
    for resource_id in resource_ids:
        credential = obtain(resource_id, is_interactive, validate_cb)
        if credential is None:
            print(f'[{mark}: NO ACCESS] no valid credential for {resource_id}, skipping it')


# Demo of "Scenario 1: non-interactive":
# res_1 is accessed
# res_2 is skipped because credential for res_2 is invalid
# res_3 is skipped because there is no credential
credential_demo(['res_1', 'res_2', 'res_3'], False, validate_access)

# Demo of "Scenario 2: interactive, credential unknown
credential_demo(['res_3'], True, validate_access)

# Demo of "Scenario 2: interactive, credential known, credential is good
credential_demo(['res_1'], True, validate_access)

# Demo of "Scenario 2: interactive, credential known, credential fails to give access
credential_demo(['res_2'], True, validate_access)

```

**Comment by christian-monch** [comment]
Another observation: credentials are requested multiple times on directory-based redirects. For example, a request URL might be

```
https://example.com/data/study_1
```

If the local path resolves to a directory, the HTTP-server might redirect it to:
```
https://example.com/data/study_1/
```

In this case the credential system asks two times for the credentials. I think it should be enough to ask only once.

**Comment by christian-monch** [comment]
Since we have settled on delegation of credential validation to the credential manager (IIRC), I thought a little bit about that yesterday evening and I would opt for implementing delegation via objects that have a `validate`-callable and encapsulate all knowledge about the validation process. Those can then be handed to `CredentialManager.obtain`.

A common implementation would be to use instances of a Validator-class (more precisely one of the subclasses). Something like:

```python

class CredentialValidator:
    def validate(location: str, credential: Credential) -> bool:
        pass

        
class HttpUrlCredentialValidator(CredentialValidator):
    def validate(location, credential):
        # Validate access here
        uops = HttpUrlOperations()
        try:
            uops.stat(location, credential)
        except UrlOperationsAuthenticationError:
            return False
        return True
```
WDYT?

**Comment by mih** [comment]
I think this is more complicated than needed: It would be `HttpUrlOperations` that employs `CredentialManager` to perform the credential workshop. If credential manager requires a `CredentialValidator` to validate a credential, this object needs to have a reference to the outer `HttpUrlOperations` inside -- otherwise it would need to duplicate an instance (as shown above), which is not generally possible or easy (think persistent sessions, etc).

So instead of `HttpUrlCredentialValidator`, this `validate()` method could be bound to the `HttpUrlOperations` instance directly, or rather be pre-parameterized with `location` (`CredentialManager` will never change the location, or even know it beyond a `realm` property).

That being said, this brings us to what I wrote above:

> I am considering whether the API of obtain() could be extended to receive a a custom callable (provided by the caller) that obtain() calls with the retrieved credential, and it returns a boolean that indicates whether this credential is fit for the intended purpose (e.g., login worked).

Conceptually the same, but with the implementation differences I described in here.

Key question is: what would be the advantage of a class with a defined API over a simple callable that takes a credential, and returns a bool?

**Comment by christian-monch** [comment]
> Conceptually the same, but with the implementation differences I described in here.
Yes.

> Key question is: what would be the advantage of a class with a defined API over a simple callable that takes a credential, and returns a bool?

A class instance could carry additional information, that might be required for successful validation in a relatively simple way. That might not be necessary though. I am probably not sufficiently aware of all the contexts in which `obtain()` is called.

I can happily agree with a callable that returns a `bool`.

**Comment by mih** [comment]
I think having a class instance and only passing a callable are not mutually exclusive. The callable can easily be an instance method.

**Comment by mih** [comment]
Started looking into this further. Returning a bool is likely not sufficient. We also need error messaging, i.e. why a credential was rejected or failed to work.

My current thinking is to have an argument `tester` which takes a callable that returns `None` if all works, or a `str` message with information on why not (service unavailable, user unknown, password wrong, ...)

**Comment by mih** [comment]
So what does this mean re implementation complexity for `CreddentialManager` users?

They would need to provide a callable that, in the simplest case, takes a single credential AND can be called more than once, in case a credential is not fit for purpose. The callable needs to run `None` on success, an error message on recoverable credential failure, and raise an exception on non-recoverable, non-credential-related failure.

This is manageable. However, the key question is: Is this complexity worth the effort given the alternative. This alternative is a `CredentialManager` method that is returning an iterator yielding any number of matching credentials. `CredentialManager` user would then need to implement the looping on their own, and for each use case. In contrast to having to implement a callable than can be called repeatedly.

In both cases the argument given to that `CredentialManager` method must be sufficient to determine number and order of matching credentials.

I feel like implementing the looping once and in `CredentialManager` has the potential for implementing the best UX, because everything there is to know about credential handling is present in a single place.

**Comment by mih** [comment]
Current concept

```py
    def call_with_credential(
               self,
               callable: Callable,
               *
               name: str | None = None,
               prompt: str | None = None,
               type_hint: str | None = None,
               query_props: Dict | None = None,
               expected_props: List | Tuple | None = None):
```

**Comment by mih** [comment]
One more UX issue: when credentials are immediately tested for validity there is no point in repeating secret entry. Either it is correct or not, typing things twice does not add anything, but is annoying.

**Comment by mih** [comment]
It seems better to have a dedicated exception to communicate an invalid credential. This gives us the ability to possibly have a series of derived exceptions that all communicate slightly different to better match difficult use cases.

**Comment by mih** [comment]
I have a basic implementation of a credential workflow. Needs lots of polishing still. Will push when I have a rich set of tests for it

**Comment by mih** [comment]
It seems like a good and well used first use case for exposing the credential workflow is `DataladAuth`, the `requests` HTTP authentication helper. The "tester" callable would be `_authenticated_rerequest()`. It would need a bit of code shuffling, but at first glance looks quite doable.
